Fix RandomPath to respect update call

kren1 opened the pull request:
This patch fixes RandomPathSearcher to respect the calls to `update` method and therefore supports use cases where you only want it to work on subset on states. It uses pointer bit stealing(`llvm::PointerIntPair`), so it's limited to just 3 instances/subsets.

Because currently the `update` method is a noop, this is of course slower, but I never saw it in the flamegraphs. The `selectState` should have similar performance as before as this patch adds only a couple of arithmetic instructions.

This should deprecate #762 and in general enable subset searcher extensions.

I had to revert @251 `unqiue_ptr` patch as it is very difficult/annoying/unclear to have bit-stolen pointers with `unqiue_ptr`.

251 commented:
As you know, I have a similar mechanism in my branch and just by skimming the code: why do you need `size`? Isn't it enough to check that both pointers are un-tagged in the corresponding bits in `empty()`?

kren1 commented:
Yeah I guess that could work too, but I like having size.

251 commented on lib/Core/Searcher.h:
```diff
@@ -166,11 +167,27 @@ namespace klee {
-    Executor &executor;
+    PTree &processTree;
+    // Number of states handled by this searcher
+    int size = 0;
```
`int` as state counter supports less than a month KLEE run-time on COTS hardware for many benchmarks...

kren1 replied:
I strongly doubt you can find hardware, where KLEE will have 2 billion states, but sure.

251 replied:
That's just ~600GB of `ExecutionState`s... at least make it `uint32_t`.

@MartinNowack I have an idea for new server hardware. :sunglasses:
https://www.dell.com/en-uk/work/shop/povw/poweredge-r940
```
Maximum RAM
RDIMM 3TB
LRDIMM 6TB 
NVDIMM 384GB
DCPMM 12.28TB (15.36 with LRDIMM)
```

kren1 replied:
wait, that's only 300B per state? with 3GB limit, it's quite good to see 100 000 states. That's 30KB per state. For 2 billion states, that's 60TB, so still too small, but quite close actually :O

251 replied:
A single `ExecutionState` object is a little bit above 300 bytes I think. This does not include all the other bookkeeping data (caches, constraints, stacks...).

kren1 replied:
I think my 30KB estimate is more realistic.

251 replied:
For Coreutils it's between 13kB and 1.2MB in my last experiment - assuming that all of the memory is used for state handling.

251 commented on lib/Core/Searcher.h:
```diff
@@ -166,11 +167,27 @@ namespace klee {
 
   public:
-    RandomPathSearcher(Executor &_executor);
+    RandomPathSearcher(PTree &_processTree);
```
Please get rid of that underscore nonsense.

251 commented on lib/Core/Executor.h:
```diff
@@ -491,6 +490,8 @@ class Executor : public Interpreter {
     symPathWriter = tsw;
   }
 
+  PTree &getPTree() { return *processTree; }
```
Getters that only return values are an antipattern. By returning a reference it becomes "public" anyway.

kren1 replied:
Would const reference be ok?

251 commented on lib/Core/Searcher.cpp:
```diff
@@ -287,12 +295,70 @@ void
                            const std::vector<ExecutionState *> &removedStates) {
-}
+  size = size + addedStates.size() - removedStates.size();
+  if (current != nullptr) {
```
Instead of checking `current` you could just inherit tags in `attach()`. The last argument is always the parent/current state - but we would need to enforce/document this.

251 commented on lib/Core/Searcher.cpp:
```diff
@@ -254,29 +254,37 @@ bool WeightedRandomSearcher::empty() {
 }
+// Check if n is a valid pointer and a node belonging to us
+#define IS_OUR_NODE_VALID(n)                                                   \
+  ((((n).getInt() & idBitMask) != 0) && ((n).getPointer() != nullptr))
```
I'd change the order: if it's null, don't check the mask.

kren1 commented:
I've been using this for a while and looking at FlameGraphs. The `RandomPathSearcher::update` doesn't show up, because I don't think it's called frequently enough to be a performance issue. `RandomPathSearcher::selectState` is quite hot, but it's cost is dominated by traversing memory, so some minor arithmetic operations don't impact it much. I think the performance impact of this patch is negligible.

251 commented on lib/Core/ExecutionState.cpp:
```diff
@@ -109,7 +109,7 @@ ExecutionState::ExecutionState(const ExecutionState& state):
     forkDisabled(state.forkDisabled),
     coveredLines(state.coveredLines),
-    ptreeNode(state.ptreeNode),
+    ptreeNode(nullptr), //ptreeNode shouldn't be shared between 2 states
```
If you don't copy just default-initialise the `ptreeNode` to `nullptr` in `ExecutionState.h`.

251 commented on lib/Core/PTree.h:
```diff
@@ -28,15 +33,28 @@ namespace klee {
 
   class PTree {
+    // Number of registered ID
+    int registeredIds = -1;
```
Meh... this semantics is ugly. How do I register a negative amount?

kren1 replied:
What are you suggesting? It's just 0 semantics, that's usually cleaner.

To answer your questions: you can't, therefore there aren't any

251 replied:
I'd expect a small unsigned type initialised with 0...

kren1 replied:
But then the shift has an ugly `1 << (registeredIds + 1)`

251 replied:
```
uint8_t id = 1 << registeredIds++;
...
return id;
```

251 commented on lib/Core/Searcher.cpp:
```diff
@@ -258,29 +258,37 @@ bool WeightedRandomSearcher::empty() {
 RandomPathSearcher::~RandomPathSearcher() {
 }
+// Check if n is a valid pointer and a node belonging to us
+#define IS_OUR_NODE_VALID(n)                                                   \
```
"OUR"?

kren1 replied:
The one we arw considering right now. That's just how I thought about it

251 replied:
`CHILD_NODE`, `IS_CHILD_NODE`, `ACCESSIBLE_CHILD_NODE`, ...

kren1 replied:
But it obviously is child node. The point is it is a node  that belongs to our subset of states and it is valid.

251 replied:
It might be a child node in PTree but not a child node for RandomPath. Maybe sth. in the direction of `IS_MAINTAINED_CHILD`...

kren1 replied:
I don't think MAINTAINED is a good word here. In my mind the searcher doesn't maintain anything. A node belongs to the searcher, ie. is part of the searcher or not.

251 replied:
Besides that a node doesn't belong to a searcher... It is just ASSIGNED to or ASSOCIATED with it.

MartinNowack replied:
Do we really need a macro here? A simple static function would do the same.

251 commented on lib/Core/Searcher.cpp:
```diff
@@ -291,12 +299,46 @@ void
                            const std::vector<ExecutionState *> &removedStates) {
-}
+  for (auto es : addedStates) {
+    PTreeNode *pnode = es->ptreeNode, *parent = pnode->parent;
```
Just `node` is too readable? ;-)

kren1 replied:
Yes

251 commented on lib/Core/Searcher.h:
```diff
@@ -13,6 +13,7 @@
 #include "klee/Internal/System/Time.h"
 
 #include "llvm/Support/CommandLine.h"
+#include "PTree.h"
```
Among LLVM includes?!

ccadar commented:
@kren1 thanks for this patch.   I would like to have a better high-level understanding of its goal, given that it complicates the code quite a lot.  

Is this mainly meant as a replacement for #762 or do you have other use cases in mind?  Are those use cases already possible in KLEE?

What do you mean by being limited to 3 subsets?  How bad is this limitation?

251 commented:
> have a better high-level understanding of its goal

All searchers but RandomPath are able to maintain a subset of states. This PR adds this subset capability to RandomPath.

> Is this mainly meant as a replacement for #762

There are many use cases, this is just on of them. Some of our projects already use this or similar implementations (also #1207 seems related).

> Are those use cases already possible in KLEE?

Besides merging? No. But we could upstream some projects.

> What do you mean by being limited to 3 subsets?

This PR annotates pointers (bit stealing, fat pointers) and can only re-use three bits per address without any further assumptions about the host architecture.

> How bad is this limitation?

Depends. How many use cases can you imagine that need more than 3 RandomPath searchers that maintain different subsets of states? I've used only 2 up to now.

kren1 commented:
In addition to what @251 said, there are 4 projects that need this:

* state merging ( use case already upstream and this replaces #762 )
* pending states (normal vs pending states subsets)
* replaying (normal vs replayed states)
* chopper (normal vs recovery states)

> given that it complicates the code quite a lot.

I would argue that it simplifies the code. It enables and hopefully enforces the separation of Searchers and Executor. Currently the code is in this weird state where everything goes through the Searcher API, except random path which has this weird exception. So wherever you are doing something general with the searchers you have to special case for RandomPath, many projects don't bother (see merging).

> What do you mean by being limited to 3 subsets? How bad is this limitation?

I can't think of a case where you  need more than 2, but I am sure having more subsets is a possibility. This should be quite easily extensible to around 10+ searchers if you are making assumptions about the architecture and also steal the high bits, not just the low ones. The more general limitation is that this PR doesn't support an arbitrary number of searcher, but a fixed small amount. This is so that performance is kept equivalent to original implementation, which is fast.

ccadar commented:
@kren1 I finally managed to properly look at this patch, and it's very nice, thank you!
Let's try to merge it soon.  I only have a few small requests:
1. It would be nice to also add a comment in PTree.h as to why `PTreeNodePtr` is needed.  It can first look confusing without knowing about RP search
2. Can you mention the limitation of having a maximum of 3 RPS instances in the comment you added in `Searcher.h`?
3. You use at various points "label", "tag" and "id" to denote the same thing.  Could you try to use a single term?
4. Can you use `klee_error` instead of `assert`+`abort`?
5. Can you squash the last commit into the previous one?
6. Minor: "To support this RP " -> "To support this, RandomPathSearcher"

kren1 commented:
@ccadar Thanks, I think I addressed your comments.

I now only use "tag", instead of "tag" and "label". I tag and ID are not the same. Each RP-Searcher has an ID (0,1,2). Tags are what is associated with the PTreeNode and store the IDs (for example they can be 0b000, 0b101 (stores IDs 0 and 2), ...). I don't see any occurrence where I use tag for id or vice versa.

ccadar commented:
Thanks, @kren1 , perfect!  Or almost perfect ;)  Now that you clarified the terms ID and tag, I'd like to ask you for a couple of small things.  I'll add them as inline comments, it's easier.  Thanks again!
(btw, it's great that you took the time to write the unit tests, they're useful)

251 commented on lib/Core/PTree.h:
```diff
@@ -11,15 +11,26 @@
   class ExecutionState;
+  class PTreeNode;
+  // PTreeNodePtr is used by the Random Path Searcher object to efficiently
+  // record which PTreeNode belong to it. PTree is a global structre that
```
PTreeNode_s_ or belong_s_

251 replied:
+ Can you check if the comments are shown in doxygen? IIRC the `//` comments are left out...

kren1 commented on lib/Core/PTree.cpp:
```diff
@@ -57,21 +66,23 @@ void PTree::dump(llvm::raw_ostream &os) {
-      stack.push_back(n->left.get());
+    if (n->left.getPointer()) {
+      os << "\tn" << n << " -> n" << n->left.getPointer();
+      os << " [tag=" << (int)n->left.getInt() << "];\n";
```
Done that, but I renamed `tag` back to `label` here. This  output is meant to be in `dot` format and the `label=` actually annotates the arrows if you plot it.

ccadar replied:
OK, makes sense then

kren1 commented:
These unit tests are surprisingly easy to write and having more of them would encourage better modularisation of KLEE. I was actually concerned how #1246 would impact these tests, but I can only say that it improved the includes.

251 commented on lib/Core/PTree.h:
```diff
@@ -28,14 +39,26 @@ namespace klee {
     static void remove(PTreeNode *node);
     void dump(llvm::raw_ostream &os);
+    uint8_t getNextId() {
+      auto id = 1 << registeredIds++;
```
Why `auto`? Doesn't it default to `int` here?

kren1 replied:
Because I didn't want to think about what is the appropriate type. Isn't that the point of `auto` in this case? I think I was hopping it would infer the type from the return type of the function.

251 replied:
No, this is not Haskell. C++ just sees the 1 and chooses `int`. You could use `decltype` here but @ccadar might hate it. :stuck_out_tongue: Just use `std::uint8_t`...

kren1 replied:
Well clearly the compiler needs to get better!

251 replied:
Or the C++ standard... However, you can declare the return type as auto: `auto getNextId() {...}`. This is allowed since C++14. :grin:

kren1 commented:
Travis seems to fail because `sqlite.org` is not reachable at the moment. Why are we even downloading sqlite in CI?

ccadar commented:
@kren1 Can you propose a better way to deal with sqlite?

kren1 commented:
Now I see this is only a problem for `*San`s, because they need the source code. We could have the sqlite-amalganation.zip checked in git, but that only really avoids this connectivity issue.

ccadar commented:
@kren1 W/o looking at the build scripts, why are the *San targets treating this differently?

kren1 commented:
I think the idea is to instrument as much code as possible, it is not necessary for ASan and UBsan, but MSan needs it (because otherwise sqlite memory is not tracked).

ccadar commented:
@kren1 I see -- so SqLite is compiled with sanitization support too.  I agree it's likely needed for MSan in order to get reasonable results.

251 commented:
> We could have the sqlite-amalganation.zip checked in git

I use the amalgamation files in MoKlee. It has the advantage that we can optimise the builds and remove unnecessary features but on the other hand we have to "maintain" it. Not the worst idea imho...

kren1 commented:
I think the trade-offs are quite minor and I guess this is the first time I've seen this connection issue, so maybe easiest to keep as is.

MartinNowack commented on unittests/Searcher/SearcherTest.cpp:
```diff
@@ -0,0 +1,217 @@
+//===-- SolverTest.cpp ----------------------------------------------------===//
```
I have an idea from which test this file originated ;)

ccadar commented:
@kren1 can you please rebase this to incorporate the recent CI fixes?

ccadar commented:
Thanks again, @kren1 , great change!
